sdk/core/azure-core/src/main/java/com/azure/core/util/ClientOptions.java

```diff
@@ -0,0 +1,127 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+package com.azure.core.util;
+
+import com.azure.core.annotation.Fluent;
+
+import java.util.Locale;
+import java.util.Map;
+import java.util.Objects;
+import java.util.concurrent.ConcurrentHashMap;
+
+/**
+ * Client Options for setting common properties for example applicationId. Most of these properties are applied on
+ * request being send to Azure Service but some could be used for other purpose also for example applicationId could be
+ * used for telemetry.
+ */
+@Fluent
+public final class ClientOptions {
+    private final Map<String, Header> headers = new ConcurrentHashMap<>();
+    private String applicationId;
+
+    /**
+     * Gets the applicationId.
+     *
+     * @return The applicationId.
+     */
+    public String getApplicationId() {
+        return applicationId;
+    }
+
+    /**
+     * Sets the applicationId provided.
+     *
+     * @param applicationId to be set.
+     * @return updated {@link ClientOptions}.
+     *
+     * @throws NullPointerException if {@code applicationId} is null.
```

Comment:
Why can't it be null? It is by default null so it seems reasonable for a user to null it out if they don't want it set any longer?